### PR TITLE
I166

### PR DESCRIPTION
**Pull request by sufftea**

### CODE REVIEW

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 1]
```diff
@@ -31,23 +32,23 @@ import 'package:solid_lints/src/models/solid_lint_rule.dart';
...
 
-  /// Creates a new instance of [AvoidDebugPrint]
+  /// Creates a new instance of [AvoidDebugPrintInReleaseRule]
   /// based on the lint configuration.
-  factory AvoidDebugPrint.createRule(CustomLintConfigs configs) {
+  factory AvoidDebugPrintInReleaseRule.createRule(CustomLintConfigs configs) {
     final rule = RuleConfig(
       configs: configs,
       name: lintName,
       problemMessage: (_) => "Avoid using 'debugPrint'",
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:48]: We need to update the problem message to let the user know what the real problem is.

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 2]
```diff
@@ -95,9 +97,28 @@ class AvoidDebugPrint extends SolidLintRule {
...
+                    AvoidDebugPrintInReleaseCheckModel.parseExpression(
+                  node.expression,
+                );
+
+                return check.isNotRelease;
+              }
+
+              return false;
+            },
+          ) !=
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:117]: Please, extract this into variable to make it more readable.

File: lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  static const String _kReleaseMode = 'kReleaseMode';
+
+  /// Whether the expression has the same source library as kDebugMode
+  bool get hasTheSameSource => _expectedPath == sourcePath;
+
+  /// Whether the expression has the same name as kReleaseMode
+  bool get hasSameName => _kReleaseMode == name;
+
+  /// Whether the expression checks ifthe mode is not release.
+  bool get isNotRelease => negated && hasSameName && hasTheSameSource;
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart:65]: Please, place these mambers before constructors/factories

File: lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+  @override
+  String toString() {
+    return '$name, $sourcePath';
+  }
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart:70]: Do we actually need this override? If so, all the methods should be declared after all constructors/factories.

File: lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  static const String _expectedPath =
+      'package:flutter/src/foundation/constants.dart';
+
+  static const String _kReleaseMode = 'kReleaseMode';
+
+  /// Whether the expression has the same source library as kDebugMode
+  bool get hasTheSameSource => _expectedPath == sourcePath;
+
+  /// Whether the expression has the same name as kReleaseMode
+  bool get hasSameName => _kReleaseMode == name;
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart:62]:
```suggestion
  bool get _isKReleaseModeName => _kReleaseMode == name;
```
I think this name would be more clear.

File: lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    );
+  }
+
+  static const String _expectedPath =
+      'package:flutter/src/foundation/constants.dart';
+
+  static const String _kReleaseMode = 'kReleaseMode';
+
+  /// Whether the expression has the same source library as kDebugMode
+  bool get hasTheSameSource => _expectedPath == sourcePath;
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart:59]:
```suggestion
  bool get _isFromFoundationConstants => _expectedPath == sourcePath;
```
?

File: lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+  /// Whether the expression has the same name as kReleaseMode
+  bool get hasSameName => _kReleaseMode == name;
+
+  /// Whether the expression checks ifthe mode is not release.
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_debug_model.dart:64]:
```suggestion
  /// Whether the expression checks if the mode is not release.
```

File: lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_func_model.dart
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/models/avoid_debug_print_in_release_func_model.dart:1]: This class looks really similar to the avoid_debug_print_in_release_debug_model.dart. We should avoid duplicating our code.

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 8]
```diff
@@ -31,23 +32,23 @@ import 'package:solid_lints/src/models/solid_lint_rule.dart';
 /// ```
 ///
 ///
-class AvoidDebugPrint extends SolidLintRule {
+class AvoidDebugPrintInReleaseRule extends SolidLintRule {
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:35]: Please, update documentation above.

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
Well done, @sufftea! I've added a couple of suggestions here - please, take a look.

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 9]
```diff
@@ -128,4 +143,63 @@ class AvoidDebugPrint extends SolidLintRule {
       reporter: reporter,
     );
   }
+
+  bool _isDebugPrintNode(Identifier node) {
+    const printPath = 'package:flutter/src/foundation/print.dart';
+    const debugPrint = 'debugPrint';
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:149]: Does it make sense to place these constants on the class level?

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 10]
```diff
@@ -128,4 +143,63 @@ class AvoidDebugPrint extends SolidLintRule {
...
+        sourcePath = node.staticElement?.librarySource?.uri.toString() ?? '';
+
+      default:
+        return false;
+    }
+
+    return name == debugPrint && sourcePath == printPath;
+  }
+
+  bool _isDebugCheck(Expression node) {
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:170]:
```suggestion
  bool _isNotReleaseCheck(Expression node) {
```
?

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 11]
```diff
@@ -128,4 +143,63 @@ class AvoidDebugPrint extends SolidLintRule {
...
+      return token.type == TokenType.BANG &&
+          _isReleaseModeIdentifier(identifier);
+    }
+
+    return false;
+  }
+
+  bool _isReleaseModeIdentifier(Identifier node) {
+    const String expectedPath = 'package:flutter/src/foundation/constants.dart';
+    const String kReleaseModeName = 'kReleaseMode';
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:185]: Wouldn't it be great to move it to class level?

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]

**Comment by solid-vovabeloded** [comment]
Also, please, update the CHANGELOG.md file with the changes you made.

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart (hunk 11 above)
Comment (sufftea) [reply lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:185]: They're only used locally though. Still move?

File: CHANGELOG.md [hunk 12]
```diff
@@ -6,7 +6,8 @@
 - Add a rule prefer_guard_clause for reversing nested if statements (https://github.com/solid-software/solid_lints/issues/91)
 - add exclude params support to avoid_returning_widgets rule (https://github.com/solid-software/solid_lints/issues/131)
 - add quick fix to avoid_final_with_getter (https://github.com/solid-software/solid_lints/pull/164)
-
+- Renamed `avoid_debug_print` to `avoid_debug_print_in_release`
+- Now, `avoid_debug_print_in_release` doesn't trigger if the `debugPrint` call is wrapped in a `!kReleaseMode` check.
```
Comment (solid-vovabeloded) [inline CHANGELOG.md:10]:
```suggestion
- The `avoid_debug_print_in_release` rule no longer reports a warning if the `debugPrint` call is wrapped in a `!kReleaseMode` check.
```

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 13]
```diff
@@ -1,13 +1,13 @@
 import 'package:analyzer/dart/ast/ast.dart';
 import 'package:analyzer/dart/ast/syntactic_entity.dart';
+import 'package:analyzer/dart/ast/token.dart';
 import 'package:analyzer/error/listener.dart';
 import 'package:custom_lint_builder/custom_lint_builder.dart';
-import 'package:solid_lints/src/lints/avoid_debug_print/models/avoid_debug_print_func_model.dart';
 import 'package:solid_lints/src/models/rule_config.dart';
 import 'package:solid_lints/src/models/solid_lint_rule.dart';
 
 /// A `avoid_debug_print` rule which forbids calling or referencing
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:9]:
```suggestion
/// An `avoid_debug_print_in_release` rule which forbids calling or referencing
```

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 14]
```diff
@@ -1,13 +1,13 @@
...
-/// debugPrint function from flutter/foundation.
+/// debugPrint function in release mode.
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:10]:
```suggestion
/// debugPrint function from flutter/foundation in release mode.
```

File: lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart [hunk 15]
```diff
@@ -25,29 +25,35 @@ import 'package:solid_lints/src/models/solid_lint_rule.dart';
...
+
+  /// Creates a new instance of [AvoidDebugPrintInReleaseRule]
   /// based on the lint configuration.
-  factory AvoidDebugPrint.createRule(CustomLintConfigs configs) {
+  factory AvoidDebugPrintInReleaseRule.createRule(CustomLintConfigs configs) {
     final rule = RuleConfig(
       configs: configs,
       name: lintName,
-      problemMessage: (_) => "Avoid using 'debugPrint'",
+      problemMessage: (_) => "Avoid using 'debugPrint' in release mode",
```
Comment (solid-vovabeloded) [inline lib/src/lints/avoid_debug_print_in_release/avoid_debug_print_in_release_rule.dart:53]:
```suggestion
      problemMessage: (_) => "Avoid using 'debugPrint' in release mode. Wrap your debugPrint call into !kReleaseMode check",
```

**Review by solid-vovabeloded — CHANGES_REQUESTED** [review]
A few more wording improvements and you'll be good to go. Great job!

**Review by solid-vovabeloded — APPROVED** [review]

**Pull request merged**